### PR TITLE
Added OpenGL version of DemoRenderer.

### PR DESCRIPTION
**Pull request by shin1m**
Hi,

I have added a couple of projects for OpenGL:
* DemoRenderer.GL
* Demos.GL

GLSL shaders are converted from HLSL by using [ShaderConductor](https://github.com/microsoft/ShaderConductor).
DemoRenderer.GL/hlsl2glsl is for reproducing what I did.

`dotnet run` in Demos.GL to run it. I have tested it on my Linux x86 box.

Please check it out when you have time.

Thank you,

### CODE REVIEW

File: Demos/GameLoop.cs [hunk 1]
```diff
@@ -23,7 +23,17 @@ public GameLoop(Window window)
...
+            Surface = new RenderSurface(
+#if OPENGL
+                window.WindowInfo,
+#else
+                window.Handle,
+#endif
+                window.Resolution
+#if DEBUG
+                , enableDeviceDebugLayer: true
+#endif
```
Comment (RossNordby) [inline Demos/GameLoop.cs:35]: This will cause crashes on windows for people who don't have the necessary developer SDKs installed when they try to run in debug mode (as I found out :P). Should remain false.

**Review by RossNordby — CHANGES_REQUESTED** [review]
Impressive work! GL path works on my machine too. Looks minimally invasive on the DX path.

Other than the debug layer issue mentioned earlier, the only issue I notice is that the rasterized shape types all have a flipped light direction. Not exactly a big problem.

Stepping up to the metalevel, this kind of large contribution is one I *usually* would be very hesitant to accept. I wouldn't usually want the extra implied burden of supporting a target I spend little time on with a partially unshared codebase, but this is a bit of a special case.

1. I don't plan on making any significant changes to the *existing* demos renderer, so maintenance effort should remain low. I suspect I'll never even need to regenerate the GLSL.
2. The reason why I don't plan to make any significant changes to the existing demos renderer is... that it will likely be completely replaced in the medium term (maybe long term, based on how my timelines usually go). My internal projects use a cross platform dx12/vulkan API abstraction and the renderer built on it could suit the demos quite well (once I get around to filling out some more of the features and get other internal projects up and running/released).

So in other words, after some number of months, I'll probably be completely replacing all rendering in the demos. That makes it a lot easier to accept this contribution, though I imagine that could be a little frustrating after spending time making this.

In summary, this looks good and I'll accept it after the couple of little changes, but I'd encourage everyone working on significant PRs in the future to get in touch early since I might be doing something internally that could step on your toes, or the PR might not be something I want to merge in the end.

**Comment by shin1m** [comment]
Thank you for the comments.

> Other than the debug layer issue mentioned earlier, the only issue I notice is that the rasterized shape types all have a flipped light direction. Not exactly a big problem.

Oh, I have not noticed it. I would try to fix it.

> Stepping up to the metalevel, this kind of large contribution is one I _usually_ would be very hesitant to accept. I wouldn't usually want the extra implied burden of supporting a target I spend little time on with a partially unshared codebase, but this is a bit of a special case.

I understand your situation. I was half prepared to be refused.

> So in other words, after some number of months, I'll probably be completely replacing all rendering in the demos. That makes it a lot easier to accept this contribution, though I imagine that could be a little frustrating after spending time making this.

That's great! Don't mind about my time. I'm looking forward to see it :-)

> In summary, this looks good and I'll accept it after the couple of little changes, but I'd encourage everyone working on significant PRs in the future to get in touch early since I might be doing something internally that could step on your toes, or the PR might not be something I want to merge in the end.

I'm sorry for my sudden PR without asking about your plan. I'm 100% agree with you.
Again, thank you for your kind response. I'll fix those you pointed out.

Regards,

**Comment by shin1m** [comment]
I have committed a couple of changes.
Please check it out when you have time.

Regards,

**Comment by RossNordby** [comment]
Looks good! Thanks for the contribution!

**Comment by shin1m** [comment]
Thank you, too!
